packages/berry-pnpify/sources/index.ts

```diff
@@ -0,0 +1,176 @@
+  const dirs: { [dir: string]: string[] } = {};
+  const locator = pnp.findPackageLocator(issuer);
+  if (!locator) {
+    return undefined;
+  }
+  const info = pnp.getPackageInformation(locator);
+  if (info.packageLocation + '/' !== issuer) {
+    return undefined;
+  }
+  const packages: string[] = Array.from(info.packageDependencies.keys());
+  packages.forEach((pkg: string) => {
+    dirs[''] = dirs[''] || [];
+    if (pkg[0] === '@') {
+      const [ scopeName, packageName ] = pkg.split('/');
+      dirs[scopeName] = dirs[scopeName] || [];
+      dirs[scopeName].push(packageName);
+      if (dirs[''].indexOf(scopeName) < 0) {
+        dirs[''].push(scopeName);
+      }
+    } else {
+      dirs[''].push(pkg);
+    }
+  });
+  return dirs[dir];
+};
+
+function getIssuerForNodeModulesPath(pnp: any, pathname: string) {
+  const locator = pnp.findPackageLocator(pathname);
+  if (!locator) {
+    return undefined;
+  }
+  let issuer = pnp.getPackageInformation(locator).packageLocation;
+  const remainder = pathname.substring(issuer.length);
+  const issuers = [issuer].concat(remainder.split(/\/node_modules/));
+
+  issuers[0] += '/';
+  issuer = issuers[0];
+  for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+    issuers[issuerIdx] = issuers[issuerIdx].substring(1);
+  }
+
+  try {
+    for (let issuerIdx = 1; issuerIdx < issuers.length; issuerIdx++) {
+      if (issuers[issuerIdx] === '') continue;
+      if (issuers[issuerIdx] === '@types') {
+        return issuer;
+      }
+      const nextIssuer = pnp.resolveToUnqualified(issuers[issuerIdx], issuer);
+      if (!nextIssuer) {
+        return undefined;
+      }
+      issuer = nextIssuer + '/';
+    }
+    return issuer;
+  } catch (e) {
+  }
+}
+
+const pnpApiRoots: { [pnpRootDir: string]: any } = {};
+const nonPnpPaths: { [pnpRootDir: string]: boolean } = {};
+
+function getPnpApiForPath(pathname: string) {
+  const basePath = pathname.substring(0, pathname.indexOf('node_modules') - 1);
+  if (nonPnpPaths[basePath]) {
+    return undefined;
+  }
+  for (let knownPnpRootDir of Object.keys(pnpApiRoots)) {
+    if (basePath.startsWith(knownPnpRootDir)) {
+      return pnpApiRoots[knownPnpRootDir];
+    }
+  }
+
+  let remainderPath = basePath;
+  while (true) {
+    const pnpApiPath = path.join(remainderPath + '/', '.pnp.js');
+    try {
+      if (fs.existsSync(pnpApiPath)) {
+        pnpApiRoots[remainderPath] = require(pnpApiPath);
+        return pnpApiRoots[remainderPath];
+      }
+    } catch (e) {}
+    const idx = remainderPath.lastIndexOf('/');
+    if (idx < 0) {
+      break;
+    }
+    remainderPath = remainderPath.substring(0, idx);
+  }
+  nonPnpPaths[basePath] = true;
+  return undefined;
+}
+
+function resolvePath(pathname: string) {
+  let targetPath = pathname;
+  const idx = pathname.lastIndexOf('node_modules');
+  const pnp = idx >= 0 && getPnpApiForPath(pathname);
+  if (pnp) {
+    const issuer = getIssuerForNodeModulesPath(pnp, pathname);
```

Comment:
I'd advise to encapsulate the `substring` into this function and make it return a `{request, issuer}` object (or `null` if  it's not recognized as a PnP path).